File: 누적 합 부수기/PRG_92344.java
import java.util.*;

class Solution {
    public static int N, M;
    
    public int solution(int[][] board, int[][] skill) {
        N = board.length;
        M = board[0].length;
        
        int[][] prefixSum = new int[N + 1][M + 1];
        int len = skill.length;
        for(int i=0; i<len; i++) {
            int[] t = skill[i];
            
            int x1 = t[1];
            int y1 = t[2];
            int x2 = t[3];
            int y2 = t[4];
            int degree = t[5];
            
            if(t[0] == 1) {
                degree *= -1;
            }
            
            // 좌표 범위 표시
            prefixSum[x1][y1] += degree;
            
            // 반대되는 값을 마킹함으로써 0을 만들어냄.
            // 누적 합 계산할 때 이 뒤부터는 0으로 나오게 된다.
            prefixSum[x2 + 1][y1] -= degree;
            prefixSum[x1][y2 + 1] -= degree;
            
            
            prefixSum[x2 + 1][y2 + 1] += degree;
        }
        
        
        // 상하
        for(int i=0; i<=M; i++) {
            for(int j=0; j<N; j++) {
                prefixSum[j + 1][i] += prefixSum[j][i];
            }
        }
        
        // 좌우
        for(int i=0; i<=N; i++) {
            for(int j=0; j<M; j++) {
                prefixSum[i][j + 1] += prefixSum[i][j];
            }
        }
        
        int ans = 0;
        for(int i=0; i<N; i++) {
            for(int j=0; j<M; j++) {
                board[i][j] += prefixSum[i][j];
                
                if(board[i][j] > 0) ans++;
            }
        }
        
        
        return ans;
    }
}

File: 누적 합 부수기/PRG_92344.kt

class Solution {
    fun solution(board: Array<IntArray>, skill: Array<IntArray>): Int {
        var N = board.size
        var M = board[0].size
        
        var len = skill.size
        
        // 누적 합
        val prefixSum = Array(N + 1) { IntArray(M + 1)}
        for(i in 0 until len) {
            val temp = skill[i]
            
            val type1 = temp[0]
            val x1 = temp[1]
            val y1 = temp[2]
            val x2 = temp[3]
            val y2 = temp[4]
            var deg = temp[5]
            
            if(type1 == 1) {
                deg *= -1
            }
            
            // 누적합 표시
            prefixSum[x1][y1] += deg
            prefixSum[x1][y2 + 1] -= deg
            prefixSum[x2 + 1][y1] -= deg
            prefixSum[x2 + 1][y2 + 1] += deg
        }
        

        
        // 누적합 계산하기
        // 상하
        for(i in 0..M) {
            for(j in 0 until N) {
                prefixSum[j + 1][i] += prefixSum[j][i]
            }
        }

        
        // 좌우        
        for(i in 0..N) {
            for(j in 0 until M) {
              prefixSum[i][j + 1] += prefixSum[i][j]  
            }
        }
    
        
        var ans = 0
        // 원래 보드랑 합치기
        for(i in 0 until N) {
            for(j in 0 until M) {
                board[i][j] += prefixSum[i][j]
                
                if(board[i][j] >= 1) ans++
            }
        }
        
        return ans
    } // End of solution()
} // End of Solution class